android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormPresenter.kt

```diff
@@ -0,0 +1,271 @@
+
+		interactor.submit(thing, currentRecordThing, fields, isDraft, { recordThing ->
+			currentRecordThing = recordThing
+
+			if (!isDraft) {
+				formInstance.ddmStructure.successPage?.let {
+					view.showSuccessPage(formInstance.ddmStructure.successPage)
+				} ?: run {
+					view.showSuccessMessage()
+				}
+				view.isSubmitEnabled(true)
+				view.sendCustomEvent(FormEvents.SUBMIT_SUCCESS, thing)
+			}
+		}, { exception ->
+			LiferayLogger.e(exception.message)
+
+			if (!isDraft) {
+				view.isSubmitEnabled(true)
+				view.showErrorMessage(exception)
+				view.sendCustomEvent(FormEvents.SUBMIT_FAILED, thing)
+			}
+		})
+	}
+
+	override fun syncFormInstance(thing: Thing, formInstance: FormInstance) {
+		isSyncing = true
+		view.showModalSyncFormLoading()
+
+		val fields = formInstance.ddmStructure.fields
+
+		if (formInstance.hasDataProvider) {
+			fetchDataProviders(thing, fields) {
+				fetchLatestDraft(thing, fields) {
+					onCompleteFetch(thing, formInstance, fields)
+				}
+			}
+		} else {
+			fetchLatestDraft(thing, fields) {
+				onCompleteFetch(thing, formInstance, fields)
+			}
+		}
+	}
+
+	override fun uploadFile(
+		thing: Thing, field: DocumentField, inputStream: InputStream, onSuccess: (DocumentRemoteFile) -> Unit,
+		onError: (Exception) -> Unit) {
+
+		interactor.uploadFile(thing, field, inputStream, onSuccess, onError)
+	}
+
+	private fun fetchDataProviders(thing: Thing, fields: MutableList<Field<*>>, onComplete: (() -> Unit)?) {
+		view.showModalEvaluateContextLoading()
+
+		interactor.evaluateContext(thing, fields, {
+			val formContext = FormContext.converter(it)
+			view.updatePageEnabled(formContext)
+
+			updateFields(formContext, fields)
+
+			onComplete?.invoke()
+		}, {
+			LiferayLogger.e(it.message)
+
+			view.hideModalLoading()
+			view.showErrorMessage(it)
+
+			onComplete?.invoke()
+		})
+	}
+
+	private fun fetchLatestDraft(thing: Thing, fields: MutableList<Field<*>>, onComplete: (() -> Unit)?) {
+		interactor.fetchLatestDraft(thing, {
+			view.hideModalLoading()
+
+			currentRecordThing = it
+
+			formInstanceRecord?.let { formInstanceRecord ->
+				updateFields(formInstanceRecord.fieldValues, fields)
+			}
+
+			onComplete?.invoke()
+		}, {
+			LiferayLogger.e(it.message)
+
+			view.hideModalLoading()
+
+			onComplete?.invoke()
+		})
+	}
+
+	private fun onCompleteFetch(thing: Thing, formInstance: FormInstance, fields: MutableList<Field<*>>) {
+		if (formInstance.isEvaluable) {
+			evaluateContext(thing, fields) {
+				isSyncing = false
+			}
+		} else {
+			isSyncing = false
+		}
+	}
+
+	private fun setOptions(fieldContext: FieldContext, optionsField: OptionsField<*>) {
+		val availableOptions = fieldContext.options as? List<Map<String, String>>
+
+		availableOptions?.let {
+			optionsField.availableOptions = ArrayList(availableOptions.map { Option(it) })
+		}
+	}
+
+	private fun setValue(fieldContext: FieldContext, field: Field<*>) {
+		if (fieldContext.isValueChanged == true) {
```

Comment:
why the == true?

Reply:
Good question hahaha, it's an optional boolean. That's why we have to do something like this, but I'll do it in a better way.